consensus/propagation/have_wants.go

```diff
@@ -282,6 +370,16 @@ func (blockProp *Reactor) handleRecoveryPart(peer p2p.ID, part *proptypes.Recove
 		return
 	}
 
+	if p != nil {
+		// FIXME investigate why the nil peer happens
+		p.DecreaseRequestCount(1)
+		select {
+		case <-blockProp.ctx.Done():
+			return
+		case p.receivedPart <- struct{}{}:
```

Comment:
we have two separate mutexes here that we're assuming are being updated at the same time but they're not since the receive routine is reading from the channel later than the decrement

probably fine 99.99% of the time, but could also cause weird bugs

Reply:
yes, the request count is decreased first to ensure when the part notification is sent, nothing weird happens and the next request is queued

Reply:
> nothing weird happens and the next request is queued

:thinking:  what happens if we receive two parts in quick succession, lowering the request count by 2 and filling the queue twice? how will we know to drain the extra request from the channel?

would it be possible to simplify this by only decrementing after something is pulled from the channel?

```go
for {
		select {
		case <-blockProp.ctx.Done():
			return
		case <-receivedPart:
		        p.DecreaseRequestCount(1)
		case req, ok := <-ps.requestChan:
...
```

Reply:
this is a very good point. Adding this:

```
case <-receivedPart:
p.DecreaseRequestCount(1)
```

is helpful to drain the `receivedPart` channel if we're not congested. SO will definitely add this to the main select.

However, we need to add the check-up so that we don't consume a part before we have bandwidth, and we should block until we receive a part. Here is a fix: d1fe33e4f16d6646da3364b012705aaf64f788bb

Reply:
do we need to decrement the request counter after receiving reading from the recieved part channel